File: src/lib/resumeWorkspace.ts
import { invoke } from "./tauri";

export type ResumeModuleKey = "summary" | "projects" | "experience" | "skills";
export type ResumeWorkspaceStepKey = "original" | "diagnosis" | ResumeModuleKey | "final";

export interface ResumeDiagnosisSection {
  assessment: string;
  missing_info: string[];
  suggestions: string[];
}

export interface ResumeDiagnosis {
  overall_summary: string;
  summary: ResumeDiagnosisSection;
  projects: ResumeDiagnosisSection;
  experience: ResumeDiagnosisSection;
  skills: ResumeDiagnosisSection;
  next_steps: string[];
}

export interface ResumeBasicProfile {
  name: string;
  gender: string;
  birth_or_age: string;
  education: string;
  phone: string;
  email: string;
}

export interface ResumeWorkspaceModuleDraft {
  input: string;
  followup_input: string;
  candidate?: string | null;
  notes: string[];
  checklist: string[];
  confirmed?: string | null;
  updated_at?: string | null;
}

export interface ResumeWorkspaceDraft {
  source_mode: "text" | "file";
  original_resume_text: string;
  original_resume_file?: string | null;
  basic_profile: ResumeBasicProfile;
  context_text: string;
  diagnosis?: ResumeDiagnosis | null;
  summary: ResumeWorkspaceModuleDraft;
  projects: ResumeWorkspaceModuleDraft;
  experience: ResumeWorkspaceModuleDraft;
  skills: ResumeWorkspaceModuleDraft;
  final_resume_text?: string | null;
  updated_at?: string | null;
}

export interface ResumeWorkspaceMeta {
  id: string;
  title: string;
  source_mode: "text" | "file" | string;
  source_name: string;
  created_at: string;
  updated_at: string;
  has_diagnosis: boolean;
  confirmed_modules: number;
  has_final_resume: boolean;
}

export interface ResumeWorkspaceState {
  active_workspace_id: string;
  workspaces: ResumeWorkspaceMeta[];
  draft: ResumeWorkspaceDraft;
}

export interface CreateResumeWorkspaceRequest {
  title?: string | null;
  source_mode?: "text" | "file" | null;
  original_resume_text?: string | null;
  original_resume_file?: string | null;
  basic_profile?: ResumeBasicProfile | null;
  context_text?: string | null;
}

export interface ResumeModuleCandidate {
  module: ResumeModuleKey;
  candidate: string;
  notes: string[];
  checklist: string[];
}

export interface ResumeWorkspaceNavItem {
  key: ResumeWorkspaceStepKey;
  label: string;
  status: "idle" | "active" | "done";
}

export const RESUME_MODULES: Array<{ key: ResumeModuleKey; label: string; placeholder: string }> = [
  { key: "summary", label: "个人简介", placeholder: "补充你的个人定位、求职方向、核心亮点、量化成果等。" },
  { key: "projects", label: "项目经历", placeholder: "补充项目背景、你的角色、技术栈、交付物、指标结果和难点。" },
  { key: "experience", label: "工作经历", placeholder: "补充职责、业务场景、协作范围、影响结果和晋升/带人情况。" },
  { key: "skills", label: "技能清单", placeholder: "补充技术栈、熟练度、使用场景、工具链和方法论。" },
];

function createEmptyModule(): ResumeWorkspaceModuleDraft {
  return {
    input: "",
    followup_input: "",
    candidate: null,
    notes: [],
    checklist: [],
    confirmed: null,
    updated_at: null,
  };
}

export function createDefaultResumeWorkspaceDraft(): ResumeWorkspaceDraft {
  return {
    source_mode: "text",
    original_resume_text: "",
    original_resume_file: null,
    basic_profile: {
      name: "",
      gender: "",
      birth_or_age: "",
      education: "",
      phone: "",
      email: "",
    },
    context_text: "",
    diagnosis: null,
    summary: createEmptyModule(),
    projects: createEmptyModule(),
    experience: createEmptyModule(),
    skills: createEmptyModule(),
    final_resume_text: null,
    updated_at: null,
  };
}

export function getModuleLabel(module: ResumeModuleKey): string {
  return RESUME_MODULES.find((item) => item.key === module)?.label ?? module;
}

export async function getResumeWorkspaceState(): Promise<ResumeWorkspaceState> {
  return invoke<ResumeWorkspaceState>("get_resume_workspace_state");
}

export async function createResumeWorkspace(request: CreateResumeWorkspaceRequest = {}): Promise<ResumeWorkspaceState> {
  return invoke<ResumeWorkspaceState>("create_resume_workspace", { request });
}

export async function switchResumeWorkspace(workspaceId: string): Promise<ResumeWorkspaceState> {
  return invoke<ResumeWorkspaceState>("switch_resume_workspace", { workspaceId });
}

export async function renameResumeWorkspace(workspaceId: string, title: string): Promise<ResumeWorkspaceState> {
  return invoke<ResumeWorkspaceState>("rename_resume_workspace", { workspaceId, title });
}

export async function deleteResumeWorkspace(workspaceId: string): Promise<ResumeWorkspaceState> {
  return invoke<ResumeWorkspaceState>("delete_resume_workspace", { workspaceId });
}

export async function saveResumeWorkspaceDraft(workspaceId: string, draft: ResumeWorkspaceDraft): Promise<ResumeWorkspaceState> {
  return invoke<ResumeWorkspaceState>("save_resume_workspace_draft", { workspaceId, draft });
}

export async function diagnoseResumeWorkspace(request: {
  resume_text: string;
  context_text?: string | null;
  resume_files?: string | null;
}): Promise<ResumeDiagnosis> {
  return invoke<ResumeDiagnosis>("diagnose_resume_workspace", { request });
}

export async function rewriteResumeWorkspaceModule(request: {
  module: ResumeModuleKey;
  resume_text: string;
  context_text?: string | null;
  resume_files?: string | null;
  module_input: string;
  confirmed_summary?: string | null;
  confirmed_projects?: string | null;
  confirmed_experience?: string | null;
  confirmed_skills?: string | null;
}): Promise<ResumeModuleCandidate> {
  return invoke<ResumeModuleCandidate>("rewrite_resume_workspace_module", { request });
}

export async function assembleResumeWorkspace(draft: ResumeWorkspaceDraft): Promise<ResumeWorkspaceDraft> {
  return invoke<ResumeWorkspaceDraft>("assemble_resume_workspace", { draft });
}

export async function exportResumeWorkspacePdf(draft: ResumeWorkspaceDraft, title?: string): Promise<string> {
  return invoke<string>("export_resume_workspace_pdf", { draft, title: title ?? null });
}

File: src-tauri/src/resume_workspace.rs
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
#[cfg(test)]
use serde_json::Value;
#[cfg(test)]
use time::format_description::well_known::Rfc3339;

use crate::storage;

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ResumeDiagnosisSection {
  #[serde(default)]
  pub assessment: String,
  #[serde(default)]
  pub missing_info: Vec<String>,
  #[serde(default)]
  pub suggestions: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ResumeDiagnosis {
  #[serde(default)]
  pub overall_summary: String,
  #[serde(default)]
  pub summary: ResumeDiagnosisSection,
  #[serde(default)]
  pub projects: ResumeDiagnosisSection,
  #[serde(default)]
  pub experience: ResumeDiagnosisSection,
  #[serde(default)]
  pub skills: ResumeDiagnosisSection,
  #[serde(default)]
  pub next_steps: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ResumeBasicProfile {
  #[serde(default)]
  pub name: String,
  #[serde(default)]
  pub gender: String,
  #[serde(default)]
  pub birth_or_age: String,
  #[serde(default)]
  pub education: String,
  #[serde(default)]
  pub phone: String,
  #[serde(default)]
  pub email: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ResumeWorkspaceModuleDraft {
  #[serde(default)]
  pub input: String,
  #[serde(default)]
  pub followup_input: String,
  #[serde(default)]
  pub candidate: Option<String>,
  #[serde(default)]
  pub notes: Vec<String>,
  #[serde(default)]
  pub checklist: Vec<String>,
  #[serde(default)]
  pub confirmed: Option<String>,
  #[serde(default)]
  pub updated_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResumeWorkspaceDraft {
  #[serde(default = "default_source_mode")]
  pub source_mode: String,
  #[serde(default)]
  pub original_resume_text: String,
  #[serde(default)]
  pub original_resume_file: Option<String>,
  #[serde(default)]
  pub basic_profile: ResumeBasicProfile,
  #[serde(default)]
  pub context_text: String,
  #[serde(default)]
  pub diagnosis: Option<ResumeDiagnosis>,
  #[serde(default)]
  pub summary: ResumeWorkspaceModuleDraft,
  #[serde(default)]
  pub projects: ResumeWorkspaceModuleDraft,
  #[serde(default)]
  pub experience: ResumeWorkspaceModuleDraft,
  #[serde(default)]
  pub skills: ResumeWorkspaceModuleDraft,
  #[serde(default)]
  pub final_resume_text: Option<String>,
  #[serde(default)]
  pub updated_at: Option<String>,
}

impl Default for ResumeWorkspaceDraft {
  fn default() -> Self {
    Self {
      source_mode: default_source_mode(),
      original_resume_text: String::new(),
      original_resume_file: None,
      basic_profile: ResumeBasicProfile::default(),
      context_text: String::new(),
      diagnosis: None,
      summary: ResumeWorkspaceModuleDraft::default(),
      projects: ResumeWorkspaceModuleDraft::default(),
      experience: ResumeWorkspaceModuleDraft::default(),
      skills: ResumeWorkspaceModuleDraft::default(),
      final_resume_text: None,
      updated_at: None,
    }
  }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiagnoseResumeRequest {
  pub resume_text: String,
  #[serde(default)]
  pub context_text: Option<String>,
  #[serde(default)]
  pub resume_files: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RewriteResumeModuleRequest {
  pub module: String,
  pub resume_text: String,
  #[serde(default)]
  pub context_text: Option<String>,
  #[serde(default)]
  pub resume_files: Option<String>,
  pub module_input: String,
  #[serde(default)]
  pub confirmed_summary: Option<String>,
  #[serde(default)]
  pub confirmed_projects: Option<String>,
  #[serde(default)]
  pub confirmed_experience: Option<String>,
  #[serde(default)]
  pub confirmed_skills: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ResumeModuleCandidate {
  #[serde(default)]
  pub module: String,
  #[serde(default)]
  pub candidate: String,
  #[serde(default)]
  pub notes: Vec<String>,
  #[serde(default)]
  pub checklist: Vec<String>,
}

fn default_source_mode() -> String {
  "text".to_string()
}

#[cfg(test)]
fn now_rfc3339() -> String {
  time::OffsetDateTime::now_utc().format(&Rfc3339).unwrap()
}

pub fn draft_path(app_data_dir: &Path) -> PathBuf {
  storage::storage_dir(app_data_dir).join("resume-workspace.json")
}

pub fn load_draft(app_data_dir: &Path) -> Result<ResumeWorkspaceDraft, String> {
  let Some(raw) = storage::read_json(&draft_path(app_data_dir)).map_err(|e| e.to_string())? else {
    return Ok(ResumeWorkspaceDraft::default());
  };
  serde_json::from_value(raw).map_err(|e| format!("解析简历工作区草稿失败：{e}"))
}

#[cfg(test)]
pub fn save_draft(app_data_dir: &Path, mut draft: ResumeWorkspaceDraft) -> Result<ResumeWorkspaceDraft, String> {
  draft.updated_at = Some(now_rfc3339());
  let value: Value = serde_json::to_value(&draft).map_err(|e| e.to_string())?;
  storage::write_json(&draft_path(app_data_dir), &value).map_err(|e| e.to_string())?;
  Ok(draft)
}

pub fn assemble_resume_text(draft: &ResumeWorkspaceDraft) -> String {
  let mut parts: Vec<String> = Vec::new();

  push_section(&mut parts, "个人简介", draft.summary.confirmed.as_deref());
  push_section(&mut parts, "项目经历", draft.projects.confirmed.as_deref());
  push_section(&mut parts, "工作经历", draft.experience.confirmed.as_deref());
  push_section(&mut parts, "技能清单", draft.skills.confirmed.as_deref());

  parts.join("\n\n")
}

fn push_section(parts: &mut Vec<String>, title: &str, content: Option<&str>) {
  let Some(content) = content.map(str::trim).filter(|it| !it.is_empty()) else {
    return;
  };
  parts.push(format!("## {title}\n{content}"));
}
